dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +247,365 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
-                    winrt::com_array<winrt::IBackgroundCondition> conditions = details.GetConditions();
-                    for (auto condition : conditions)
                     {
-                        builder.AddCondition(condition);
+                        auto lock{ m_lock.lock_exclusive() };
+                        builder5.SetTaskEntryPointClsid(m_registeredClsid);
+                        m_pushTriggerRegistration = builder.Register();
                     }
+
                 }
-            }
 
-            BackgroundTaskRegistration registeredTaskFromBuilder{ nullptr };
+                // Register a sink for the application to receive foreground raw notifications
+                RegisterForegroundSinkHelper();
+
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    // Register a PushNotificationBackgroundTask to handle background activation scenarios
+                    THROW_IF_FAILED(::CoRegisterClassObject(
+                        m_registeredClsid,
+                        winrt::make<PushNotificationBackgroundTaskFactory>().get(),
+                        CLSCTX_LOCAL_SERVER,
+                        REGCLS_MULTIPLEUSE,
+                        &m_comActivatorRegistration));
+                }
 
-            auto scopeExitToCleanRegistrations{ wil::scope_exit(
-                [&]()
+                scopeExitToCleanRegistrations.release();
+            }
+            else
+            {
                 {
-                    s_comActivatorRegistration.reset();
+                    auto lock{ m_lock.lock_shared() };
+                    THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_ALREADY_REGISTERED), m_singletonForegroundRegistration && m_singletonBackgroundRegistration);
+                }
 
-                    // Clean the task registration only if it was created during this call
-                    if (registeredTaskFromBuilder)
+                auto scopeExitToCleanRegistrations{ wil::scope_exit([&]()
+                {
                     {
-                        registeredTaskFromBuilder.Unregister(true);
+                        auto lock { m_lock.lock_exclusive() };
+
+                        m_comActivatorRegistration.reset();
+                        m_singletonBackgroundRegistration = false;
+                        m_singletonForegroundRegistration = false;
+                        m_registering = false;
                     }
-                }
-            ) };
 
-            if (WI_IsFlagSet(registrationActivators, PushNotificationRegistrationActivators::ComActivator))
-            {
-                GUID taskClsid = details.TaskClsid();
-                THROW_HR_IF(E_INVALIDARG, taskClsid == GUID_NULL);
+                    auto notificationsLongRunningPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                  
+                    LOG_IF_FAILED(notificationsLongRunningPlatform->UnregisterForegroundActivator(m_processName.get()));
```

Comment:
locks?

Reply:
https://github.com/microsoft/WindowsAppSDK/pull/2122#discussion_r814358341

Let me know what you think.